### PR TITLE
Add decoder

### PR DESCRIPTION
**Pull request by marktoda**
Two commits - first one adds a linter and runs it. The second one is the meat of the PR. I suggest reviewing commit-by-commit to simplify it a bit.

```
Add eslint and run lint-fix

```

```
Add contract call decoder

This commit adds a class that takes in the raw bytes of a contract call
and returns a readable explanation of the function call that it
represents. This is helpful for many situations where one needs to
determine the intent of a transaction when they only have the raw
transaction bytes.
```

### CODE REVIEW

**Comment by marktoda** [comment]
Here is how the decoder works:

```
▶ node
> const { Decoder } = require('.')
> const decoder = new Decoder()
> const decoded = decoder.decode(Buffer.from('a9059cbb00000000000000000000000010d4f942617a231eb1430c88fe43c8c2050437d90000000000000000000000000000000000000000000000000000000000002710', 'hex'));
{ methodId: '0xa9059cbb',
  name: 'transfer',
  args:
   [ { name: '_to',
       type: 'address',
       value: '0x10d4f942617a231eb1430c88fe43c8c2050437d9' },
     { name: '_value', type: 'uint256', value: 10000 } ],
  contractName: 'StandardERC20' }
```

File: .eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    "object-curly-spacing": ["error", "always", {"objectsInObjects": true, "arraysInObjects": true}],
+    "array-bracket-spacing": ["error", "never"],
+    "require-yield": "off",
+    "func-call-spacing": ["error", "never"],
+    "quotes": ["error", "single", { "avoidEscape": true, "allowTemplateLiterals": true }],
+    "key-spacing": ["error", { "beforeColon": false, "afterColon": true, "mode": "strict" }],
+    "quote-props": ["error", "as-needed"],
+    "no-console": "off",
+    "no-cond-assign": "error",
+    "no-empty": ["warn", { "allowEmptyCatch": false }],
```
Comment (wesgraham) [inline .eslintrc.json:37]: Do we want this to error instead? Can't think of any place where empty conditionals would be desirable

File: .eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    "no-irregular-whitespace": "error",
+    "no-duplicate-case": "error",
+    "@typescript-eslint/no-explicit-any": "off",
+    "@typescript-eslint/explicit-module-boundary-types": "off",
+    "@typescript-eslint/no-this-alias": "warn",
+    "@typescript-eslint/no-var-requires": "off",
+    "@typescript-eslint/no-unused-vars": ["error", {
+      "vars": "all",
+      "varsIgnorePattern": "^_",
+      /* TODO: enable this rule and fix errors
```
Comment (wesgraham) [inline .eslintrc.json:82]: Are these todo's still todo's?

File: src/decoder/index.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import { ensure } from '../util/ensure';
+import { isValidJSON } from '../util/json';
+
+/**
+ * A class to decode contract calls and explain their purpose
+ */
+export class Decoder {
+
+  /**
+     * Read in and parse all methods from all defined contract abis
```
Comment (wesgraham) [inline src/decoder/index.ts:18]: Nit: Spacing

File: src/decoder/index.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  /**
+     * Decode the given function call data, returning a readable explanation of the call
+     * @param data The data to decode
+     * @return An explanation of the call, including the function name and arguments passed
+     */
+  public decode(data: Buffer): FunctionCallExplanation {
+    const methodId = bufferToHex(data.slice(0, 4));
+    const abiEncodedArguments = data.slice(4);
```
Comment (wesgraham) [inline src/decoder/index.ts:60]: I probably am incorrect but are the abiEncodedArguments == the first 5 bytes of data?

File: src/decoder/types.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+enum Primitive {
+    Address,
+    Bool,
+    String,
+    Bytes,
+    Int,
+}
+
+function formatBool(value: Buffer): boolean {
```
Comment (wesgraham) [inline src/decoder/types.ts:12]: These all look right but should we be implementing our own primitives like this? Is there any library we can use that might be a bit cleaner?

File: .eslintrc.json (hunk 1 above)
Comment (marktoda) [reply .eslintrc.json:37]: done, I just copied this whole config from platform's

File: .eslintrc.json (hunk 2 above)
Comment (marktoda) [reply .eslintrc.json:82]: removed, idk - it was from platform's

File: src/decoder/index.ts (hunk 3 above)
Comment (marktoda) [reply src/decoder/index.ts:18]: the linter did that :laughing:

Comment (marktoda) [reply src/decoder/index.ts:18]: I can't find the rule for it to disable it right now, but the nice thing about auto-fixing linters is that we don't have to worry about style nits like this and can just update the linter and re-fix if we decide later

Comment (marktoda) [reply src/decoder/index.ts:18]: Ah - found the linter rule for this. fixed ;)

File: src/decoder/types.ts (hunk 5 above)
Comment (marktoda) [reply src/decoder/types.ts:12]: Unfortunately not. The type strings are solidity types (uint256, uint256[], etc.), and I wasn't able to find a library that does this for us. ethereumjs-abi parses and decodes the types, but decides to convert them all to `Buffer` unfortunately (except address), so the point of this file is just to convert to the native JS types that we know and love. If you have an alternative idea, let me know - I'm not too mad about having to do it ourselves here, though. For example if we decide that we want to change the way the formatting looks (i.e. remove 0x prefix, add EIP-55 casing, etc.) - we have the control and power to do it ourselves this way

Comment (wesgraham) [reply src/decoder/types.ts:12]: That's alright with me - just wanted to make sure there was some logic for it :)

**Review by wesgraham — APPROVED** [review]

**Pull request merged**